dev/src/recursive-delete.ts

```diff
@@ -0,0 +1,317 @@
+      )
+    );
+
+    // Query for names only to fetch empty snapshots.
+    query = query.select(FieldPath.documentId()).limit(limit);
+
+    if (ref instanceof CollectionReference) {
+      // To find all descendants of a collection reference, we need to use a
+      // composite filter that captures all documents that start with the
+      // collection prefix. The MIN_KEY constant represents the minimum key in
+      // this collection, and a null byte + the MIN_KEY represents the minimum
+      // key is the next possible collection.
+      const nullChar = String.fromCharCode(0);
+      const startAt = collectionId + '/' + REFERENCE_NAME_MIN_ID;
+      const endAt = collectionId + nullChar + '/' + REFERENCE_NAME_MIN_ID;
+      query = query
+        .where(FieldPath.documentId(), '>=', startAt)
+        .where(FieldPath.documentId(), '<', endAt);
+    }
+
+    if (startAfterLastSnapshot) {
+      query = query.startAfter(this.lastDocumentSnap);
+    }
+
+    return query.stream();
+  }
+
+  /**
+   * Called when all descendants of the provided reference have been streamed
+   * or if a permanent error occurs during the stream. Deletes the developer
+   * provided reference and wraps any errors that occurred.
+   * @private
+   */
+  private onQueryEnd(): void {
+    this.allDescendantsFetched = true;
+    if (this.ref instanceof DocumentReference) {
+      this.writer.delete(this.ref).catch(err => this.incrementErrorCount(err));
+    }
+    this.writer.flush().then(async () => {
+      if (this.lastError === undefined) {
+        this.completionDeferred.resolve();
+      } else {
+        let error = new GoogleError(
+          `${this.errorCount} ` +
+            `${this.errorCount !== 1 ? 'deletes' : 'delete'} ` +
+            'failed. The last delete failed with: '
+        );
+        if (this.lastError.code !== undefined) {
+          error.code = (this.lastError.code as number) as Status;
+        }
+        error = wrapError(error, this.errorStack);
+
+        // Wrap the BulkWriter error last to provide the full stack trace.
+        this.completionDeferred.reject(
+          this.lastError.stack
+            ? wrapError(error, this.lastError.stack ?? '')
+            : error
+        );
+      }
+    });
+  }
+
+  /**
+   * Deletes the provided reference and starts the next stream if conditions
+   * are met.
+   * @private
+   */
+  private deleteRef(docRef: DocumentReference): void {
+    this.pendingOpsCount++;
+    this.writer
+      .delete(docRef)
+      .catch(err => {
+        this.incrementErrorCount(err);
+      })
+      .then(() => {
+        this.pendingOpsCount--;
+        // We wait until the previous stream has ended in order to sure the
+        // startAfter document is correct. Starting the next stream while
+        // there are pending operations allows Firestore to maximize
+        // BulkWriter throughput.
+        if (
+          !this.allDescendantsFetched &&
+          !this.streamInProgress &&
```

Comment:
I wonder if could ever hit this code path after all documents are done deleting but before the stream ended. That seems unlikely, but if it is possible it would mean that recursive delete stalls. Do we need to also call `setupStream` in `onStreamEnd()`? I am not sure but you might have better judgement here.

Reply:
After our conversation, I added a `setupStream()` call at the end.